ms_active_directory/core/ad_session.py

```diff
@@ -263,6 +263,92 @@ def _create_object(self, object_dn: str, object_classes: List[str], account_attr
                                       'and object classes {}. LDAP result: {}'.format(object_dn, object_classes,
                                                                                       result))
 
+    def create_user(self, username: str, first_name: str, last_name: str, object_location: str,
+                    **additional_user_attributes) -> ADUser:
+        """ Use the session to create a user in the domain and return a user object.
+        :param username: The login username for the user to create in the AD domain. This will be used to determine
+                         the sAMAccountName for the user.
+        :param first_name: The first name (given name) of the user to create in the AD domain. This will used for the
+                           givenName attribute for the user and as a part of the user display name.
+        :param last_name: The last name (surname) of the user to create in the AD domain. This will used for the
+                          sn (surname) attribute for the user and as a part of the user display name.
+        :param object_location: The distinguished name of the location within the domain where the user will be
+                                created. It may be a relative distinguished name (not including the domain component)
+                                or a full distinguished name.  If not specified, defaults to CN=Users which is
+                                standard for Active Directory.
+        :param additional_user_attributes: Additional LDAP attributes to set on the user and their values.
+                                           This is used to support power users setting arbitrary attributes.
+                                           This also allows overriding of some values that are not explicit keyword
+                                           arguments in order to avoid over-complication, since most people won't
+                                           set them (e.g. userAccountControl).
+        :returns: an ADUser object representing the user.
+        :raises: ObjectCreationException if we fail to create the user for a reason unrelated to what we can
+                 easily validate in advance (e.g. permission issue)
+        """
+        logger.debug('Request to create user in domain %s with the following attributes: username=%s, '
+                     'first_name=%s, last_name=%s, object_location=%s, number of additional attributes specified: %s',
+                     self.domain_dns_name, first_name, last_name, object_location, len(additional_user_attributes))
+
+        if self.object_exists_in_domain_with_attribute(ldap_constants.AD_ATTRIBUTE_SAMACCOUNT_NAME, username):
+            raise ObjectCreationException('An object already exists with sAMAccountName {} so a user may not be '
+                                          'created with the username {}'.format(username, username))
```

Comment:
there's a function for validating `sAMAccountName`s that are used for logon in `ldap_utils`. it's used in the computer creation function and should be applied here as well before we bother looking up the name in the domain. it'll check things like allowable length and characters (microsoft has some restrictions for backwards compatibility)

it might also be prudent to take in a parameter around supporting legacy auth for this use, the way that computer creation does. if a user's `sAMAccountName` exceeds 15 characters, they won't be able to use NTLM properly

Reply:
This function appears to be written specifically for computer `sAMAccountName`s, would you rather I create a separate but very similar function for user `sAMAccountName`s or generalize the existing function for both users and computers?

Reply:
ah, it does the removal of the `$` from the end and the error message says computer, but the general validation it does around length, allowed characters, and ntlm is still going to be the same. so we could just update the error message and the function would work for users

Reply:
Done.